Biogenerator tweaks, leather makes more belts and clothing

MTandi opened the pull request:
## About The Pull Request

### Revamped the biogenerator UI:

https://user-images.githubusercontent.com/3625094/200973283-b703f21b-c747-493e-98d9-043eef86d410.mp4

### Changed biogenerator icon to use layers and see the biomass level:

https://user-images.githubusercontent.com/3625094/201396065-caeaa412-6676-46f6-875e-efa2dca34985.mp4

### Biogenerator rebalance:

- Now you don't need the beaker to print solid products.
- Biogenerator now accepts all food, not just plants.
- Biogenerator now treats all nutriment subtypes as nutriments, so vitamins and proteins also turn into biomass.
- Biomass now has the same units as other reagents (you get 5 biomass from 5 nutrient with tier 1 parts).
- Doubled the cost of all items and reagents. (biomass generation reduced by 10 and prices - by 5)
- Chemicals output amounts are now in units and you can select how much you want to output exactly. It will not let you specify more than the size of container or above 50 units with one button click.
- Reduced the amount of stored items and introduced a limit to the biomass, both tied to the matter bin tier.

### Recipes changes:

Made biogenerator more dumb by moving the clothing out from the biogenerator designs, and extending leather recipes instead.

The biogenerator is a grinder/recycler style machine so it doesn't make sense that it outputs clothing.
Also you need to make leather to craft the toolbelt, while you can't do the same to craft job-specific belts.
Now you can print leather in biogenerator and craft the leather clothing by using the leather in-hand.
And the rice hat is now crafted from bamboo, instead of biogenerator.

Also added paper to the biogenerator recipes as it makes stuff from cellulose and barely anyone knows that you can craft paper from 1 log and 50 water. And paper is needed in large quantities to craft some items, like paper frames.

And it doesn't output a pack of rolling paper. It's dumb now. It prints the rolling paper sheets instead.

## Why It's Good For The Game

Biogenerator had terrible UX and backend logic. I didn't improve much on BE though, but now it should be less frustrating to use.

Also I hate how biogenerator is superior to all other means of obtaining its products. It doesn't make sense to grow and grind wheat, for instance, when you can just throw shit into biogenerator and get the flour fast. And the costs are ridiculous - you can get a couple of bottles of fertilizers just from one medium potato. 

It honestly begs for more nerfing, at least to make the nutriment - chemicals exchange rate 1:1.

The reason for the biomass cap is because people use it as a sink for veggies and generate infinite biomass. Maybe the limit will make them care more about the part upgrade and offload some of the veggies to the fridge for the Cook.

Also it was weird that biogenerator could tailor some things, while others have to be crafted in-hand. Now you can print leather and craft all types of belts and leather clothing.

## Changelog
:cl:
refactor: biogenerator UI revamped
qol: biogenerator no longer requires beaker for materials, monkey cubes and nori
balance: biogenerator accepts all food, not just plants
balance: biogenerator treats all nutriment subtypes as nutriments (vitamins, protein, etc.)
balance: biogenerator product prices doubled
balance: biogenerator biomass storage is limited depending on the level of matter bins
balance: cowboy boots recipe moved from crafting to leather recipes
balance: leather clothing & belt recipes moved from biogenerator to leather recipes
balance: rice hat recipe moved from biogenerator to bamboo recipes
balance: biogenerator now outputs rolling paper sheets instead of a pack
add: biogenerator can now print paper
imageadd: biogenerator icons now use overlays, have emissive layer and indicate the biomass volume
/:cl:

MTandi committed:
no sewing in biogenerator

MTandi committed:
UI update

MTandi commented:
Sorry, I'm not familiar with the UI tech stack enough to do this.

Mothblocks commented on code/game/objects/items/stacks/sheets/leather.dm:
```diff
@@ -188,15 +188,23 @@ GLOBAL_LIST_INIT(xeno_recipes, list ( \
-	new/datum/stack_recipe("saddle", /obj/item/saddle, 5), \
-	new/datum/stack_recipe("sheriff vest", /obj/item/clothing/accessory/vest_sheriff, 4), \
+	new/datum/stack_recipe_list("belts", list( \
+		new/datum/stack_recipe("tool belt", /obj/item/storage/belt/utility, 4), \
```
Broken indentation

MTandi replied:
It is not, there is nesting, just like the cowboy hats and fancy boxes for cardboard are.

Mothblocks replied:
i see

Mothblocks commented on code/game/objects/items/stacks/sheets/leather.dm:
```diff
@@ -188,15 +188,23 @@ GLOBAL_LIST_INIT(xeno_recipes, list ( \
+		new/datum/stack_recipe("security belt", /obj/item/storage/belt/security, 2), \
+		new/datum/stack_recipe("shoulder holster", /obj/item/storage/belt/holster, 3), \
+		new/datum/stack_recipe("bandolier", /obj/item/storage/belt/bandolier, 5), \
+		)),
```
```suggestion
	)),
```

I know the other ones are like this but they shouldn't be

MTandi commented:
Marked as draft as I ended up rewriting UI to tsx and making the logic better while I'm at it. (Like not needing a beaker to print solid objects)

MTandi committed:
draft ts UI

MTandi commented:
> Did you manage to fix the weird thing where the generator needs a beaker in it to grind?

Yes

MTandi committed:
Biogenerator UI revamp

MTandi committed:
indent fix

Wallemations commented:
Oh wait did you remove the reagent color for the beaker's bar? That has 10 units but is the default blue, what did you fill it with?

Admittedly I probably should have taken the reagent color and mixed it with the default blue so even when it changes it and its text are easily visible

MTandi committed:
amount tweaks

Wallemations commented:
> > Oh wait did you remove the reagent color for the beaker's bar? That has 10 units but is the default blue, what did you fill it with?
> > Admittedly I probably should have taken the reagent color and mixed it with the default blue so even when it changes it and its text are easily visible
> 
> I didn't notice it had such functionality at first. Noticed this variable and added to the new version. Please check out the video.

That looks great! I didn't know there was a way to add shadows to the text, so that also fixes another issue I had. Nice job!

MTandi commented:
> One thing that I'm not so big on is the biomass cap, since getting a higher one would mean someone in science would have to go around with stock parts, which in my experience is rare. That's more a thing for an official maintainer to talk to you about though.

We have resource cap for almost all machines and the purpose of matter bins is to increase it, so this is more of a consistency fix. 

You can also go to science by yourself to ask for the parts just like you go to Chemistry when you need Saltpetre or Diethylamine. This is actually another positive thing about this change - more department interaction reasons.

Biogenerator should not be special, and the main reason for the cap is that it could hold an infinite amount of biomass and sometimes botanists grind crops that are useful for the cook, without even thinking.

MTandi committed:
max reagents limited only by beaker

MTandi committed:
max can't be 0

MTandi committed:
Smooth biomass generation

MTandi committed:
oopsie

MTandi committed:
better sound

MTandi committed:
fix

MTandi committed:
Merge branch 'tgstation:master' into biosewing

MTandi committed:
doubled item capacity

MTandi committed:
Merge branch 'biosewing' of github.com:MTandi/tgstation into biosewing

MTandi commented:
>The second thing is unless the above is happening, we also set aside a high nutriment crop as a “biomass” producer. This is important because we run through biomass pretty fast when we don’t have chem dispensers or plumbing for more powerful fertilizer.

Well, the cap won't affect you because you said that you run through the biomass pretty fast. It just shouldn't be infinite. You can try to build from this branch and suggest to tweak the values. The main idea is to make the basic setup good enough for the beginning, but also make the upgraded parts desired. 

And it's your job to upgrade your tools. Go ask science for parts, if they ignore this, you have a valid reason to tide them. You could find out that there is no resources in lathe because miners also don't do their job, and hydroponics could also run out of power if engineers don't do their job. The chemist can tell you to GTFO with your request for saltpetre. This game is about cooperation, you shouldn't expect to have a pleasant experience when the whole shift is in a chaos. 

Everyone depends on engineering for power, on chemistry for chemicals, on science for parts and on security for defending from antags.

DustanHache commented:
> The price change will also not be a problem for you if you keep high nutriment crops for biomass. They provide way more biomass than you will need even with the basic parts. You shouldn't just dump a harvest from one potent tomato plant and get fertilizers for the rest of the shift.
>
> And it's your job to upgrade your tools. Go ask science for parts, if they ignore this, you have a valid reason to tide them. You could find out that there is no resources in lathe because miners also don't do their job, and hydroponics could also run out of power if engineers don't do their job. The chemist can tell you to GTFO with your request for saltpetre. This game is about cooperation, you shouldn't expect to have a pleasant experience when the whole shift is in a chaos.
> 
> Everyone depends on engineering for power, on chemistry for chemicals, on science for parts and on security for defending from antags. You should remember about the pleasant experience when the science guys actually come and upgrade your stuff without you asking, and we want to have these. They wouldn't bother if it was insignificant for you.

MTandi commented:
@DustanHache I just tested how it works with the standard seed from vendor and e-z nutrient. 4 harvests from the same tomato plant gave me 267 biomass equal to 2670 units or 38 buckets of e-z nutrient. Or 890 units of Enduro grow.

![image](https://user-images.githubusercontent.com/3625094/201345196-355014bc-ef93-4975-8dbd-c740fcf001a9.png)

And watermelons will give you double. So, I don't really get the concern.

MTandi committed:
insert beaker while processing

MTandi commented:
@DustanHache what are your suggestions?

DustanHache commented:
> @DustanHache what are your suggestions?

Double the cap, retain the original prices for fertilizer (roughly, Decimals are finnicky) and that should be sufficient to keep a mildly competent botany from dying the moment a chef wants cheese.

MTandi committed:
new animations

MTandi committed:
icon fixes

MTandi commented:
> Double the cap, retain the original prices for fertilizer (roughly, Decimals are finnicky) and that should be sufficient to keep a mildly competent botany from dying the moment a chef wants cheese.

I already adjusted the cap and numbers slightly since the beginning.
* Do you think 500 biomass is not enough for level 1 parts? Better 1000 at level 1 and 4000 at level 4?
* Do you think that 10 fertilizer units per 1 unit of e-z nutrient is not enough? Should it be 20 or 25 like before?

Please assume that this is not the only source of fertilizers, but there is also the vendor with free bottles.

MTandi committed:
update icon with part upgrade

DustanHache commented:
> I already adjusted the cap and numbers slightly since the beginning.
> 
> * Do you think 500 biomass is not enough for level 1 parts? Better 1000 at level 1 and 4000 at level 4? Note that the current cap with level 1 part will give you 5000 units of e-z nutrient.

Yes, but this isn’t because of a fertilizer shortage. In fact, i would argue that you could take some of the expansion from parts and “front end” it so that the end result is closer to your original numbers. This reduces how much they will inevitably be hamstrung by others not doing their jobs/refusing to cooperate whilst still allowing them to cater to the chef(s) occasional spike in demand.
> * Do you think that 10 fertilizer units per 1 unit of nutriment is not enough? Should it be 20 or 25 like before?
Yes, ~20 fertilizer per 1 biomass is the current balance and it is not something to change without good reason, such as a expansion to the reagent production capabilities. The exceptions are more exotic fertilizers (eg, endurogrow) and the pesticides which are significantly more expensive due to their unique effects or utility in maintaining good harvests.

> Please assume that this is not the only source of fertilizers, but there is also the vendor with free bottles.
Vendors do not give free items in TG by default unless it is an off-station area such as lavaland. We get a discount instead for being of the correct department. With this in mind, we have to be careful how we spend our paychecks to ensure we don’t fry our limited budget while being forced to pay taxes for our upgrades and seeds. Ultimately the biogenerator does become the main/only stable source unless chemistry is generous or poorly protected.

MTandi committed:
balance tweaks

MTandi commented:
@DustanHache thanks for the feedback. I made the following tweaks, hope this helps:

- Initial biomass cap is doubled - its 1000 now, but the increment with part upgrades is still 500, so the max on t4 is 2500
- Initial plant storage cap is doubled - its 40 now, but the increment is 20, so it's 100 items with t4 parts
- Halved prices on all liquid reagents except pest sprays and nori, to ease the access to essential chemicals for botany and reduce the spikes when chefs print their shit.

Let's revisit these if needed after playing with it for some time.

On a side note I want to ask you as Botany main, what do you think about the biogenerator giving flour and sugar? The availability of these chemicals make wheat and sugar cane redundant as it's easier to just drop a watermelon into the reactor.
Yes, chefs don't bug botany for wheat and sugar now, but it feels like cheating. Maybe biogenerator should give some subtype of a milk that is a bit worse than the natural counterpart for cooking?

MTandi commented:
Man, I played with it. The first harvest of watermelons from the seed vendor gives 200 biomass, and it's 4000 ez nutrient. You overdramatized this a lot, but whatever now. Maybe you didn't know that you need to keep the nutrient level topped out for the best immediate results.

DustanHache commented:
> @DustanHache thanks for the feedback. I made the following tweaks, hope this helps:
> 
> * Initial biomass cap is doubled - its 1000 now, but the increment with part upgrades is still 500, so the max on t4 is 2500
> * Initial plant storage cap is doubled - its 40 now, but the increment is 20, so it's 100 items with t4 parts
> * Halved prices on all liquid reagents except pest sprays and nori, to ease the access to essential chemicals for botany and reduce the spikes when chefs print their shit.

As i said before, I’m okay with the price bump on everything that isn’t fertilizers. Chef spam will eat a ton, but that’s okay once botany is up and running at full capacity.

> On a side note I want to ask you as Botany main, what do you think about the biogenerator giving flour and sugar? The availability of these chemicals make wheat and sugar cane redundant as it's easier to just drop a watermelon into the reactor. Yes, chefs don't bug botany for wheat and sugar now, but it feels like cheating. Maybe biogenerator should give some subtype of a flour or milk that is a bit worse than the natural counterpart for cooking?

personally I like it in terms of producing flour, as it is painfully hard to get sufficient quantity from wheat. Same goes for milk and soy milk. Sugar is the main one that I disagree with, as the plants that produce it are very generous in terms of production. That said, it’s also one of the reagents that they have a lower use rate for..

> I also just dislike that the biogenerator makes protein products in form of the milk and monkey cubes from plants. Especially when there are some alternatives in form of cows and monkey powder recipe.

I feel like monkey cubes are fine where they are, as chef isn’t the only one who has use for them. That said, I personally wouldn’t mind a price increase on them either.
Milk is where things get interesting though.... Chef starts with pete, who also gives milk. The issue is you have to keep him fed, alive, and deal with any temper tantrums they throw. Also, their milk is of limited supply for the purposes of mass production, leading to issues of bottlenecking on some recipes that are milk hungry, such as cheese. With those things in mind, I would like to point to the life expectancy of the average spessman and/or pun pun and ask you: do you really think that a goat is going to survive longer unless they’re shoved in a fridge?



> Man, I played with it. The first harvest of watermelons from the seed vendor gives 200 biomass, and it's 4000 ez nutrient. You overdramatized this a lot, but whatever now. Maybe you didn't know that you need to keep the nutrient level topped out for the best immediate results.

Yes, i am aware that nutrient amounts influence how fast stat changes are applied, though i’ve had several cases where for one reason or another Botany ends up without power and without the ability to purchase more fertilizers. I am surprised at how much biomass that yields, and i feel like something isn’t adding up. Did you top off the trays before or during growth, or just plant and proceed to “wait”? I generally go get tools and hack my seed vendor so I can get glowshrooms (in case engineering fucks up) while growing my first set of plants, and don’t fertilize them beyond what starts in the tray until my first harvest (to save money for contraband seeds.)

MTandi commented:
>Yes, i am aware that nutrient amounts influence how fast stat changes are applied, though i’ve had several cases where for one reason or another Botany ends up without power and without the ability to purchase more fertilizers. I am surprised at how much biomass that yields, and i feel like something isn’t adding up. Did you top off the trays before or during growth, or just plant and proceed to “wait”? I generally go get tools and hack my seed vendor so I can get glowshrooms (in case engineering fucks up) while growing my first set of plants, and don’t fertilize them beyond what starts in the tray until my first harvest (to save money for contraband seeds.)

I used the seed from vending machine with like 20 starting potency and 3-4 yield. Before the melons matured, they already had almost 100 potency and 10 yield. The first batch gave me 10 watermelons with 20 nutrient in each. And they ate about 10 units of fertilizer. 

I used hydroponics tray with autogrow enabled and topped it up with ez nutrient every 20 seconds. There's will be an automatic way of doing this BTW https://github.com/tgstation/tgstation/pull/71217

You can just keep a watermelon stored in a fridge and use it if the chef uses up all biomass. 1 melon will give 400u of ez nutrient. And before nerf it gave 200 biomass which was equal to 500u of ez. Man I don't know how you struggled, I won't even be able to use it up fast enough.

DustanHache commented:
I used the seed from vending machine with like 20 starting potency and 3-4 yield. Before the melons matured, they already had almost 100 potency and 10 yield. The first batch gave me 10 watermelons with 20 nutrient in each. And they ate about 10 units of fertilizer.

the fertilizer usage matches up, but not the yield and potency. I’m not sure what’s going on there. Maybe random mutations?

> I used hydroponics tray with autogrow enabled and topped it up with ez nutrient every 20 seconds. There's going to be an automatic way of doing this

this should still not affect yield, meaning your harvest should be only 4. Your method also skews the results slightly, as fertilizer quantity affects how fast the stats rise, as I said before. Regarding your automation comment, that is already possible via plumbing. You just need a constructor, an input, and a tank or two to hold it your excess.
> You can just keep a watermelon stored in a fridge and use it if the chef uses up all biomass. 1 melon will give 400u of ez nutrient. And before nerf it gave 200 biomass which was equal to 500u of ez. Man I don't know how you struggled, I won't even be able to use it up fast enough.
This matches up with max potency/yield melons, and does represent the effectiveness of botany once fully operational.

MTandi committed:
bugfix

ExcessiveUseOfCobblestone commented:
The botany bounties are arguably the easiest in the game, If you are paycheck to paycheck that is because you don't know the mechanics provided to you or you're just choosing to gimp yourself.

DustanHache commented:
> The botany bounties are arguably the easiest in the game (in terms of overall accessibility to the items), If you are paycheck to paycheck that is because you don't know the mechanics provided to you or you're just choosing to gimp yourself.

It is a good moneymaker, but it relies on cargo being mentally functional beyond mining, or someone EAing the shuttle.

MTandi committed:
linting

MTandi committed:
Merge branch 'tgstation:master' into biosewing

MTandi committed:
Merge branch 'tgstation:master' into biosewing

MTandi committed:
Biomass spills on deconstruction

Mothblocks commented on code/modules/hydroponics/biogenerator.dm:
```diff
@@ -47,59 +53,78 @@
 	if(default_deconstruction_crowbar(O))
+		if(biomass > 0)
+			drop_location.visible_message(span_warning("The biomass spills!"))
+			playsound(drop_location, 'sound/effects/slosh.ogg', 25, TRUE)
```
```suggestion
			playsound(drop_location, 'sound/effects/slosh.ogg', 25, vary = TRUE)
```

Mothblocks commented on code/modules/hydroponics/biogenerator.dm:
```diff
@@ -167,74 +192,54 @@
-		use_power(processing_time * active_power_usage * 0.1) // .1 needed here to convert time (in deciseconds) to seconds such that watts * seconds = joules
-		sleep(processing_time + 15 / productivity)
+		var/potential_biomass = 0
+		for(var/obj/item/food/grown/I in contents)
```
Proper name

Mothblocks commented on code/modules/hydroponics/biogenerator.dm:
```diff
@@ -167,74 +192,54 @@
+					use_power(active_power_usage * 0.1) // .1 needed here to convert time (in deciseconds) to seconds such that watts * seconds = joules
+					potential_biomass -= 1
+					biomass += 1
+					sleep(2 / productivity)
```
```suggestion
					stoplag(2 / productivity)
```

Mothblocks commented on code/modules/hydroponics/biogenerator.dm:
```diff
@@ -167,74 +192,54 @@
+			qdel(I)
+			if(potential_biomass)
+				while(processing && potential_biomass > 0)
+					use_power(active_power_usage * 0.1) // .1 needed here to convert time (in deciseconds) to seconds such that watts * seconds = joules
```
```suggestion
					use_power(active_power_usage * (1 SECONDS)) // Seconds needed here to convert time (in deciseconds) to seconds such that watts * seconds = joules
```

MTandi replied:
@Mothblocks We fucked up here, it should've been 0.01 SECONDS. Now the Biogen uses 300kw instead of 3kw. I will make a PR to fix that. I noticed this only after it got into master.

Mothblocks commented on code/modules/hydroponics/biogenerator.dm:
```diff
@@ -167,74 +192,54 @@
+					potential_biomass -= 1
+					biomass += 1
+					sleep(2 / productivity)
+					update_appearance()
```
What actually needs updating here? If it's just the icon, change to:

```suggestion
					update_appearance(UPDATE_ICON)
```

MTandi replied:
luminosity as well

Mothblocks commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -358,6 +358,7 @@ GLOBAL_LIST_INIT(bamboo_recipes, list ( \
 	new/datum/stack_recipe("bamboo spear", /obj/item/spear/bamboospear, 25, time = 90), \
 	new/datum/stack_recipe("blow gun", /obj/item/gun/syringe/blowgun, 10, time = 70), \
 	new/datum/stack_recipe("crude syringe", /obj/item/reagent_containers/syringe/crude, 5, time = 10), \
+	new/datum/stack_recipe("rice hat", /obj/item/clothing/head/costume/rice_hat, 10, time = 70), \
```
```suggestion
	new/datum/stack_recipe("rice hat", /obj/item/clothing/head/costume/rice_hat, 10, time = 7 SECONDS), \
```

MTandi committed:
SECONDS

MTandi committed:
codereview

MTandi committed:
re-nerf liquid prices

MTandi committed:
biogenerator accepts all food now

MTandi commented:
I played with it a bit and made a few more balance changes based on the experience:

1. Reverted the price for liquids back to be higher. Before this PR you got 25 eznut per 1 nutriment, and now - 10
2. Instead, now biogenerator treats all subtypes of nutriments as nutriments too. So Pumpkins, Watermelons give 24 biomass and stop being the only good option for biomass. Sweet potato, for instance, gives 20, because it has 10 vitamins.
3. **BIOGENERATOR NOW ACCEPTS ALL FOOD**, not just plants. So the chef may process his own shit to make milk instead of begging for biomass. You can even inject organ tissue into a donk pocket and turn it all into biomass.
4. Reverted the caps to be lower. I didn't see anyone on terry making a lot of biomass anyway, and plants now have more biomass, so the plant capacity change is also justified.

If anyone still thinks that the prices for nutrients is high, he should learn the mechanics and explore other ways of obtaining fertilizers. I think that the prices should be even higher now. The biogenerator is a recycler, you should not be able to shove a few plump helmets there to keep all your hydroponics running.

MTandi committed:
Merge branch 'tgstation:master' into biosewing

MTandi committed:
use any bag
